jmake/jmkmf.man
''' $Id$
'''
'''  Copyright (c) 1991-1997, 2004-2006, Raphael Manfredi
'''  
'''  You may redistribute only under the terms of the Artistic Licence,
'''  as specified in the README file that comes with the distribution.
'''  You may reuse parts of this distribution only within the terms of
'''  that same Artistic Licence; a copy of which may be found at the root
'''  of the source tree for dist 4.0.
'''
''' $Log: jmkmf.man,v $
''' Revision 3.0.1.1  1995/05/12  11:58:03  ram
''' patch54: updated my e-mail address
'''
''' Revision 3.0  1993/08/18  12:04:20  ram
''' Baseline for dist 3.0 netwide release.
'''
.TH JMKMF 1 ram
.SH NAME
jmkmf \- runs jmake with the correct options
.SH SYNOPSIS
.B jmkmf
[
.I top-level dir
[
.I current dir
]
]
.SH DESCRIPTION
.I Jmkmf
is a wrapper which calls
.I jmake
with the correct options, defining the symbols
.SM TOPDIR
(location of the top-level directory) and
.SM CURDIR
(current directory). The generated
.I Makefile.SH
is then ran through
.I /bin/sh
to produce a
.I Makefile.
.PP
.I Jmkmf
is useful when you generate a makefile for the first time.
.PP
When ran without arguments,
.I jmkmf
will scan the directories upwards, looking for a .package file marking
the top of your sources.  It will then derive the top-level directory
and the name of the current directory by itself.
.PP
Once you have a
.I Makefile.SH
generated by
.I jmake,
and have run Configure already, you can use
.I make Makefile.SH
to build the Makefile.SH again and
.I make Makefile
to run the Makefile.SH through
.I /bin/sh.
To use the recursive commands, you have to append an 's' at the
end of the name as in
.I make Makefiles.SH
and
.I make Makefiles.
.SH AUTHOR
Raphael Manfredi <Raphael.Manfredi@pobox.com>
.SH FILES
.PD 0
.TP 20
Jmakefile
High level description of makefile
.TP
.package
The file marking the top of your package tree
.PD
.SH "SEE ALSO"
jmake(1), packinit(1).


ChangeLog
Fri Feb 28 17:48:41 MET 1997   Raphael Manfredi <Raphael_Manfredi@grenoble.hp.com>

. Description:

	This huge set of patch results from an integration of the
	perl5 units into the dist-3.0 package. A few improvements
	of mine are also incorporated here, but this patch set should
	mainly be credited to Andy Dougherty and Chip Salzenberg, who
	have both assumed the perl5 Patch Pumpkin with success and brio.

	I won't document most of the changes in detail. Only the obvious
	ones were kept here. You can look closer to the patches themselves
	to see what really changed. Especially the new units, since you
	now have more symbols to check for in your C code...

	Regenerated Configure with current set of units.

	Added Perl Patch Pumpkin credits. Thanks to Andy Dougherty and
	to Chip Salzenberg for their incredible contributions to
	metaconfig.

	Changed my email address and updated copyright dates

	Added support for src.U. This means you can now run Configure
	from a remote directory to separate the build directory and
	the (possibly read-only) source tree.
	Note that some of your private units may require some changes
	due to that...

	Removed support for NO_PROTOTYPE detection on SCO.

	Added cute quoting trick for wild stringify support, which
	allows non-ANSI stringification to occur.

	Replaced .a with $_a all over the place.
	Likewise for .o replaced by $_o.
	You'll have to do that to your own private units.

	Integrated new units from perl5. I will not document most
	of the perl5-related changes here. People interested can
	look at Andy's notes in the perl5 distribution.

	New USE_BSD_GETPGRP to detect the getpgrp() flavour. The old
	USE_BSDPGRP has been obsoleted in favor of USE_BSD_SETPGRP.
	Indeed, setpgrp() and getpgrp() can have distinct flavors,
	so we need two distinct tests for them.

	Useless units dropped: d_group.U and d_passwd.U.

	Mention that <poll.h> can be included if HAS_POLL defined.

	Improved overlapping copy check.

	Added support for Free_t, the type of free().

	Don't use nm with the GNU C library.
	Added support for Linux shared libs.

	Don't prompt them if the void support is high enough for us
	and don't ask for the include path unless they are on a MIPS.

	Added the srcdir support to "configure" now that we have src.U.

	Documented the runnning environment and the src.U unit in
	the metaconfig man page. The $_a and $_o are also mentionned.
	All the generated scripts must now start with a "$startsh"
	to ensure proper shell execution.

	New "create" and "empty" lint directives.

	New -L option for metaxref to match metaconfig and metalint.

	Documents contents of the message sent by patnotify:
	we now let them know the patch priority and description.

	Added a whole bunch of new units, mostly from the perl5 team.

. Files changed:

	* bin/packinit.SH: Fixed one wrong ':' comment in .package.

	* jmake/files/Jmake.rules: Now handles USRINC for dependencies. 
	Smarter about dependencies computation.

	* jmake/files/Jmake.tmpl: Lex path can now be configured.

	* mcon/U/Begin.U: Added Extractall dependency.

	* mcon/U/Config_sh.U, mcon/U/Oldsym.U, mcon/U/Extract.U,
	mcon/U/patchlevel.U, mcon/U/MailAuthor.U, mcon/U/Chk_MANI.U,
	mcon/U/Config_h.U: Added support for src.U.

	* mcon/U/Configdir.U: Have README explicitly mention the package
	name.

	* mcon/U/make.U, mcon/U/d_strtoul.U, mcon/U/Unix.U,
	mcon/U/d_inetaton.U, mcon/U/sitearch.U, mcon/U/d_sanemcmp.U,
	mcon/U/i_locale.U, mcon/U/d_sigaction.U, mcon/U/d_getpgid.U,
	mcon/U/src.U, mcon/U/sitelib.U, mcon/U/d_brokstat.U, mcon/U/man3dir.U,
	, mcon/U/d_sigsetjmp.U, mcon/U/i_sysstat.U, mcon/U/d_strtod.U,
	mcon/U/d_byacc.U, mcon/U/sh.U, mcon/U/Signal.U, mcon/U/d_gnulibc.U,
	mcon/U/man1dir.U, mcon/U/Extractall.U, mcon/U/i_values.U,
	mcon/U/i_sfio.U: Created.

	* mcon/U/Getfile.U: Getfile script now begins with "startsh".

	* mcon/U/Head.U: Make sure we unset CDPATH for shells that support
	this.  Improved Korn shell detection and handling.

	* mcon/U/Instruct.U: Logname / whoami sequence rewritten to use case.

	* mcon/U/Loc.U: Allow users to specify paths on the command line. 
	Will now substitute cp for ln if not supported.

	* mcon/U/Myread.U: Myread script now starts with a "startsh". 
	Miscellaneous fixes.

	* mcon/U/Oldconfig.U: Added support for src.U.  New OSNAME define. 
	Can now sense new OSes.

	* mcon/U/Options.U: Optdef.sh now starts with a "startsh".  Moved
	some code from Head.U.

	* mcon/U/Whoa.U: Whoa script now starts with leading "startsh".

	* mcon/U/afs.U: Can now explicitly tell Configure whether AFS is
	running.

	* mcon/U/archlib.U: Skip existence checks for archlib.

	* mcon/U/archname.U: Changed the way the archname is mangled from
	uname.

	* mcon/U/byteorder.U: No longer ask the user if the test runs ok.

	* mcon/U/ccflags.U: Removed support for NO_PROTOTYPE detection on
	SCO.  New locincpth variable.  Added info on the "additional ld
	flags" question.

	* mcon/U/cf_who.U: New computation method avoiding use of temporary
	file.

	* mcon/U/cpp_stuff.U: Added cute quoting trick for wild stringify
	support.

	* mcon/U/d_bsdjmp.U, mcon/U/d_fd_set.U, mcon/U/models.U,
	mcon/U/d_scannl.U, mcon/U/d_open3.U, mcon/U/d_closedir.U: Added ?F:
	metalint hint.

	* mcon/U/libs.U, mcon/U/libnm.U, mcon/U/libyacc.U, mcon/U/libflex.U,
	mcon/U/libdbm.U, mcon/U/d_socket.U, mcon/U/d_normsig.U,
	mcon/U/d_crypt.U: Replaced .a with $_a all over the place.

	* mcon/U/d_csh.U: Added full_csh to preserve the full path even when
	portable.

	* mcon/U/d_dosuid.U: Moved unit to TOP via a ?Y: layout directive. 
	Tell them /dev/fd is not about floppy disks.

	* mcon/U/d_gconvert.U: Integrated new unit from perl5.

	* mcon/U/d_getpgrp.U: New USE_BSD_GETPGRP to detect the getpgrp()
	flavour.

	* mcon/U/d_group.U, mcon/U/d_passwd.U: Useless unit dropped.

	* mcon/U/d_usendir.U, mcon/U/d_havetlib.U, mcon/U/errnolist.U:
	Replaced .a with $_a all over the place.  Likewise for .o replaced by
	$_o.


jmake/jmkmf.SH
case $CONFIG in
'')
	if test -f config.sh; then TOP=.;
	elif test -f ../config.sh; then TOP=..;
	elif test -f ../../config.sh; then TOP=../..;
	elif test -f ../../../config.sh; then TOP=../../..;
	elif test -f ../../../../config.sh; then TOP=../../../..;
	else
		echo "Can't find config.sh."; exit 1
	fi
	. $TOP/config.sh
	;;
esac
case "$0" in
*/*) cd `expr X$0 : 'X\(.*\)/'` ;;
esac
echo "Extracting jmake/jmkmf (with variable substitutions)"
$spitshell >jmkmf <<!GROK!THIS!
$startsh
!GROK!THIS!
$spitshell >>jmkmf <<'!NO!SUBS!'
# @(#) Generates a Makefile from a Jmakefile

# $Id$
#
#  Copyright (c) 1991-1997, 2004-2006, Raphael Manfredi
#  
#  You may redistribute only under the terms of the Artistic Licence,
#  as specified in the README file that comes with the distribution.
#  You may reuse parts of this distribution only within the terms of
#  that same Artistic Licence; a copy of which may be found at the root
#  of the source tree for dist 4.0.
#
# $Log: jmkmf.SH,v $
# Revision 3.0.1.2  2004/08/21 23:18:13  ram
# patch71: automatically figures the top dir and the current dir
# patch71: don't run Makefile.SH if the jmake call failed
#
# Revision 3.0.1.1  1993/08/19  06:42:14  ram
# patch1: leading config.sh searching was not aborting properly
#
# Revision 3.0  1993/08/18  12:04:19  ram
# Baseline for dist 3.0 netwide release.
#

usage="usage:  $0 [top_of_sources_pathname [current_directory]]"

curdir=

case $# in 
    0)
		if test -f .package; then topdir=.;
		elif test -f ../.package; then topdir=..;
		elif test -f ../../.package; then topdir=../..;
		elif test -f ../../../.package; then topdir=../../..;
		elif test -f ../../../../.package; then topdir=../../../..;
		else
			echo "$0: WARNING: can't determine top package directory" >&2
		fi
		;;
    1) topdir=$1 ;;
    2) topdir=$1  curdir=$2 ;;
    *) echo "$usage" 1>&2; exit 1 ;;
esac

case "$topdir" in
    -*) echo "$usage" 1>&2; exit 1 ;;
esac

case "$curdir" in
'')
	here=`pwd`
	top=`cd $topdir; pwd`
	curdir=`perl -e \
		'print substr($ARGV[0], length($ARGV[1])+1), "\n";' $here $top`
	case "$curdir" in
	'') curdir=.;;
	esac
	;;
esac

if test -f Jmakefile; then
	: ok
else
	echo "$0: no Jmakefile found in current directory" >&2
	exit 1
fi

if test -f Makefile.SH; then 
    echo mv Makefile.SH Makefile.SH~
    mv Makefile.SH Makefile.SH~
fi

args="-DTOPDIR=$topdir -DCURDIR=$curdir"

echo jmake $args
if jmake $args; then
	echo sh Makefile.SH
	sh Makefile.SH
else
	echo "jmake failed, aborting" >&2
	exit 1
fi
!NO!SUBS!
chmod 755 jmkmf
$eunicefix jmkmf
